### PR TITLE
Fix: redirections in hash router mode

### LINKED ISSUE #832: Url is mangled when using vue-router in hash mode
So, I'm using latest nhost-vue sdk.
In my app I'm working now with UI for password-reset / email verify / magic link - staff requried redirection.
My app current url looks something like
https://www.mysite.com/v2test/spa2/#/testpage
and this is the url I set in nhost "Client URL" config.
When I'm calling password reset function, I receive email with url, that got redirected to my https://www.mysite.com/v2test/spa2/#/testpage with additional part like
?refreshToken=xxx&type=passwordReset#refreshToken=xxx&type=passwordReset
but resulting url user presented with became rewritten to: 
https://www.mysite.com/#/testpage
I've spent several hours trying to understand why I'm keep getting wrong resulting url (missing "v2test/spa2" part) thinking it was something with my router. But then I tried manually open url with different parameter (like "ZefreshToken=xxx") and everything was working just fine, no sudden url rewrite. I'm 99% sure now - it's somethin nhost SDK does with url rewrite when it's redirected.

### PR DESCRIPTION
**Pull request by plmercereau**
- Will fix #832 after [this fix in hasura-auth is released](https://github.com/nhost/hasura-auth/pull/315)
- Allows using a hash as redirection, for instance `{ redirectTo: '#/my-page' }`

### CODE REVIEW

File: packages/hasura-auth-js/src/utils/url.ts [hunk 1]
```diff
@@ -88,12 +101,23 @@ export function removeParameterFromWindow(name: string) {
...
+    }
+
+    const hash = new URLSearchParams(location.hash?.slice(1))
+    /** @deprecated refresh tokens and types are not send in the hash anymore */
+    if (hash.get(name)) {
+      hash.delete(name)
+      url += `#${hash.toString()}`
+    } else {
+      if (location.hash) {
+        url += location.hash
```
Comment (szilarddoro) [inline packages/hasura-auth-js/src/utils/url.ts:117]:
```suggestion
        url = `${url}${location.hash}`
```

File: packages/hasura-auth-js/src/utils/url.ts [hunk 2]
```diff
@@ -88,12 +101,23 @@ export function removeParameterFromWindow(name: string) {
...
-    if (Array.from(hash).length) url += `#${hash.toString()}`
+    if (Array.from(search).length) {
+      url += `?${search.toString()}`
+    }
+
+    const hash = new URLSearchParams(location.hash?.slice(1))
+    /** @deprecated refresh tokens and types are not send in the hash anymore */
+    if (hash.get(name)) {
+      hash.delete(name)
+      url += `#${hash.toString()}`
```
Comment (szilarddoro) [inline packages/hasura-auth-js/src/utils/url.ts:114]:
```suggestion
      url = `${url}#${hash.toString()}`
```

File: packages/hasura-auth-js/src/utils/url.ts [hunk 3]
```diff
@@ -88,12 +101,23 @@ export function removeParameterFromWindow(name: string) {
...
   if (location) {
     const search = new URLSearchParams(location.search)
-    const hash = new URLSearchParams(location.hash?.slice(1))
     search.delete(name)
-    hash.delete(name)
     let url = window.location.pathname
-    if (Array.from(search).length) url += `?${search.toString()}`
-    if (Array.from(hash).length) url += `#${hash.toString()}`
+    if (Array.from(search).length) {
+      url += `?${search.toString()}`
```
Comment (szilarddoro) [inline packages/hasura-auth-js/src/utils/url.ts:107]:
```suggestion
      url = `${url}?${search.toString()}`
```

File: packages/hasura-auth-js/src/utils/url.ts [hunk 4]
```diff
@@ -58,17 +69,19 @@ export const rewriteRedirectTo = <T extends RedirectOption>(
   }
   return {
     ...otherOptions,
-    redirectTo: encodeQueryParameters(url.origin + pathName, combinedParams)
+    redirectTo: encodeQueryParameters(
+      url.origin + pathName,
```
Comment (szilarddoro) [inline packages/hasura-auth-js/src/utils/url.ts:73]:
```suggestion
      `${url.origin}${pathName}`,
```

File: packages/hasura-auth-js/src/utils/url.ts [hunk 5]
```diff
@@ -45,7 +49,14 @@ export const rewriteRedirectTo = <T extends RedirectOption>(
   }
   const baseClientUrl = new URL(clientUrl)
   const clientParams = Object.fromEntries(new URLSearchParams(baseClientUrl.search))
-  const url = new URL(redirectTo.startsWith('/') ? baseClientUrl.origin + redirectTo : redirectTo)
+
+  const url = new URL(
+    redirectTo.startsWith('/')
+      ? baseClientUrl.origin + redirectTo
```
Comment (szilarddoro) [inline packages/hasura-auth-js/src/utils/url.ts:55]:
```suggestion
      ? `${baseClientUrl.origin}${redirectTo}`
```

File: packages/hasura-auth-js/src/utils/url.ts [hunk 6]
```diff
@@ -13,8 +17,8 @@ export const encodeQueryParameters = (baseUrl: string, parameters?: Record<strin
         return `${key}=${encodeURIComponent(stringValue)}`
       })
       .join('&')
-  if (encodedParameters) return `${baseUrl}?${encodedParameters}`
-  else return baseUrl
+  const url = encodedParameters ? `${baseUrl}?${encodedParameters}` : baseUrl
+  return hash ? url + hash : url
```
Comment (szilarddoro) [inline packages/hasura-auth-js/src/utils/url.ts:21]:
```suggestion
  return hash ? `${url}${hash}` : url
```

**Review by szilarddoro — CHANGES_REQUESTED** [review]
I think template literals should be preferred over string concatenation as it improves readability.

**Comment by szilarddoro** [comment]
Can we extend our tests?

**Comment by plmercereau** [comment]
> Can we extend our tests?

done

**Review by szilarddoro — APPROVED** [review]

**Review by szilarddoro — APPROVED** [review]

File: .changeset/chatty-eels-search.md [hunk 7]
```diff
@@ -0,0 +1,5 @@
+---
+'@nhost/hasura-auth-js': patch
+---
+
+Preserve the redirection url hash it has no refresh token/type in it
```
Comment (nunopato) [inline .changeset/chatty-eels-search.md:5]:
```suggestion
Preserve the redirection url hash if has no refresh token/type in it
```

**Review by nunopato — APPROVED** [review]

**Pull request closed without merging**